python/mspasspy/db/database.py

```diff
@@ -574,45 +645,182 @@ def save_data(self, mspass_object, mode="promiscuous", storage_mode='gridfs', di
+                if save_schema.is_defined(k):
+                    if save_schema.readonly(k):
+                        if k in changed_key_list:
+                            newkey='READONLYERROR_'+k
+                            copied_metadata.change_key(k,newkey)
+                            mspass_object.elog.log_error('Database.save_data',
+                                'readonly attribute with key='+k+' was improperly modified.  Saved changed value with key='+newkey,
+                                ErrorSeverity.Complaint)
+                        else:
+                            copied_metadata.erase(k)
+            # Done editing, now we convert copied_metadata to a python dict
+            # using this Metadata method or the long version when in cautious or pedantic mode
+            insertion_dict = dict()
+            if mode == "promiscuous":
+                # A python dictionary can use Metadata as a constructor due to
+                # the way the bindings were defined
+                insertion_dict = dict(copied_metadata)
+            else:
+                # Other modes have to test every key and type of value
+                # before continuing.  pedantic kills data with any problems
+                # Cautious tries to fix the problem first
+                # Note many errors can be posted - one for each problem key-value pair
+                for k in copied_metadata:
+                    if save_schema.is_defined(k):
+                        if not isinstance(copied_metadata[k], save_schema.type(k)):
+                            insertion_dict[k]=copied_metadata[k]
+                        else:
+                            if mode == 'pedantic':
+                                mspass_object.kill()
+                                message ='pedantic mode error:  key='+k
+                                if not save_schema.is_defined(k):
+                                    message += ' not defined in schema.'
+                                else:
+                                    value=copied_metadata[k]
+                                    message += ' type of stored value='+str(type(value))+' does not match schema expectation='+str(save_schema.type(k))
+                                mspass_object.elog.log_error('Database.save_data',
+                                    'message',ErrorSeverity.Invalid)
+                            else:
+                                # Careful if another mode is added here.  else means cautious in this logic
+                                 try:
+                                     # The following convert the actual value in a dict to a required type.
+                                     # This is because the return of type() is the class reference.
+                                     insertion_dict[k] = save_schema.type(k)(copied_metadata[k])
+                                 except Exception as err:
+                                     mspass_object.kill()
```

Comment:
In our previous implementation, under cautious mode, we would only kill objects when we are unable to convert the mismatch required attribute. For those normal attributes, we only post a complaint elog entry to the mspass_object. Just to double confirm that the current change is what we expect right?

Reply:
I found the behavior in update_metadata is different from here. I think we need to agree on what's the correct behaviors under pedantic and cautious mode in tomorrow's meeting.

Reply:
That is a good point and indeed one worth discussing.   For our discussion I confirm my intention was:
1.  save_data behaves as before in cautious mode.   That is, it attempts to fix any type mismatch errors and only considers and unrecoverable type mismatch as very bad. Acton on 'very bad' is different here though (I think).   Now the result is a kill of mspass_error instead of the older approach of throwing an exception.   
2. The change I made for update_metadata was intentional.   An "update", by definition is different from "create" in the standard CRUD idea of database transactions.  In an update, of course, the record has to already exist and you are adding to it or replacing one or more name-value pairs.  In fact, writing this led me to make a new change I'll describe in a moment.  My decision, which I probably should have indeed discussed when I was implementing this, was to treat it differently because an update is a different thing.   In an update you don't want to destroy the original record by posting something demonstrably wrong.    In pedantic mode the logic is supposed to refuse to update any data with a schema mismatch.  Cautious and pedantic are handling type mismatches the same - they try to correct type mismatches, alway post an elog message about success of failure, and remove the key-value pair from the update record if the type cannot be converted.   The only difference between 'cautious' and 'pedantic' is in handling keys not defined in the schema.  'pedantic' drops them from the update while 'cautious' adds that data anyway.  both cases cause a message to be posted to elog but with different messages for the different actions. 

In responding I changed the method signature to this:
```
def update_metadata(self, mspass_object, collection=None, mode='cautious',
                exclude_keys=None, force_keys=None, alg_name='Database.update_metadata'):
```
with this explanation of the new force_keys argument in the docstring:
```
       :param force_keys: a list of metadata attributes you want to force
         to be updated.   Normally this method will only update attributes
         that have been marked as changed since creation of the parent data
         object.  If data with these keys is found in the mspass_object they
         will be added to the update record.
```
Definitely this who issue needs to be on our agenda today